src/neo/IO/Caching/HashSetCache.cs

```diff
@@ -0,0 +1,95 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Neo.IO.Caching
+{
+    public class HashSetCache<T> : IReadOnlyCollection<T> where T : IEquatable<T>
+    {
+        /// <summary>
+        /// Sets where the Hashes are stored
+        /// </summary>      
+        private readonly LinkedList<HashSet<T>> sets = new LinkedList<HashSet<T>>();
+
+        /// <summary>
+        /// Maximum capacity of each bucket inside each HashSet of <see cref="sets"/>.
+        /// </summary>        
+        private readonly int bucketCapacity;
+
+        /// <summary>
+        /// Maximum number of buckets for the LinkedList, meaning its maximum cardinality.
+        /// </summary>
+        private readonly int maxBucketCount;
+
+        /// <summary>
+        /// Entry count
+        /// </summary>
+        public int Count { get; private set; }
+
+        public HashSetCache(int bucketCapacity, int maxBucketCount = 10)
+        {
+            if (bucketCapacity <= 0) throw new ArgumentOutOfRangeException($"{nameof(bucketCapacity)} should be greater than 0");
+            if (maxBucketCount <= 0) throw new ArgumentOutOfRangeException($"{nameof(maxBucketCount)} should be greater than 0");
+
+            this.Count = 0;
+            this.bucketCapacity = bucketCapacity;
+            this.maxBucketCount = maxBucketCount;
+            sets.AddFirst(new HashSet<T>());
+        }
+
+        public bool Add(T item)
+        {
+            if (Contains(item)) return false;
+            Count++;
+            if (sets.First.Value.Count < bucketCapacity) return sets.First.Value.Add(item);
+            var newSet = new HashSet<T>
+            {
+                item
+            };
+            sets.AddFirst(newSet);
+            if (sets.Count > maxBucketCount)
+            {
+                Count -= sets.Last.Value.Count;
+                sets.RemoveLast();
+            }
+            return true;
+        }
+
+        public bool Contains(T item)
+        {
+            foreach (var set in sets)
+            {
+                if (set.Contains(item)) return true;
+            }
+            return false;
+        }
+
+        public void ExceptWith(IEnumerable<T> items)
+        {
+            foreach (var item in items)
+            {
+                foreach (var set in sets)
+                {
+                    if (set.Remove(item))
```

Comment:
When we add an element and it exceed the bucket count, it will remove all the bucket, but with this Remove we are removing only one (expected behaviour) this could produce that we have buckets unoptimized.

```
10
10
1
1
1
1
10
1
10
```

Reply:
Yes, I also realize that this happens, but tidying up the bucket is a matter that affects performance. Since this class is a cache, I think we will let this happen. Additionally I can make a little bit improvement. If the bucket is emptied, remove the bucket from the sets.

Reply:
Also we can enlarge the max capacity of each bucket to SentHashes and KnownHashes to reduce the effect by this case.